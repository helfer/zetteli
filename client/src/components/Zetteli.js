import React from 'react';

import EditableDateTime from './EditableDateTime';
import EditableTagList from './EditableTagList';
import EditableText from './EditableText';

export default class Zetteli extends React.Component {


    updateText = (evt) => {
       this.props.onUpdate({
           id: this.props.id,
           body: evt.target.value,
       });
    }
    
    onDelete = () => {
        this.props.onDelete(this.props.id);
    }

    render() {
        return (
          <div className="ui card centered fluid">
            <div className="extra content">
              <EditableDateTime datetime={this.props.datetime} /> 
              <span className="right floated">
                <i className="trash icon" onClick={this.onDelete} />
              </span>
              <EditableTagList tags={this.props.tags}/> 
            </div>
            <EditableText text={this.props.body} onChange={this.updateText} /> 
          </div>
        );
    }
}