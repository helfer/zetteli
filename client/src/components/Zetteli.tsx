import * as React from 'react';

import EditableDateTime from './EditableDateTime';
import EditableTagList from './EditableTagList';
import EditableText from './EditableText';

export interface Props {
    id: string;
    body: string;
    tags: string[];
    datetime: Date;
    onUpdate: (arg: {id: string, tags?: string[], body?: string}) => void;
    onDelete: (id: string) => void;
}

export default class Zetteli extends React.Component<Props, object> {
    updateText = evt => {
       this.props.onUpdate({
           id: this.props.id,
           body: evt.target.value,
       });
    }
    
    updateTags = newTags => {
       this.props.onUpdate({
           id: this.props.id,
           tags: newTags,
       });
    }

    onDelete = () => {
        this.props.onDelete(this.props.id);
    }

    render() {
        return (
          <div className="ui card centered fluid">
            <div className="extra content">
              <EditableDateTime datetime={this.props.datetime} /> 
              <span className="right floated">
                <i className="trash icon" onClick={this.onDelete} />
              </span>
              <EditableTagList tags={this.props.tags} updateTags={this.updateTags}/> 
            </div>
            <EditableText text={this.props.body} onChange={this.updateText} /> 
          </div>
        );
    }
}