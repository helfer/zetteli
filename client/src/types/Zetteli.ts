export interface ZetteliType {
    id: string,
    body: string,
    tags: string[],
    datetime: Date,
}