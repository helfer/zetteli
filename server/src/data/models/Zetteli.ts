import { Connector, ID } from '../connectors/ConnectorInterface';
import { isValidStackId } from '../../utils/random';

export interface ZetteliType {
    id: string;
    tags: string[];
    datetime: Date;
    body: string;
}

export default class Zetteli {
    constructor(private connector: Connector<ZetteliType>) {}

    get(id: ID) {
        return this.connector.get(id);
    }

    getAll(sid: string) {
        return this.connector.getAll(sid);
    }

    create(sid: string, zli: ZetteliType) {
        if (!isValidStackId(sid)) {
            throw new Error(`Invalid stack id: ${sid}`);
        }
        return this.connector.create(sid, zli);
    }

    update(zli: ZetteliType) {
        return this.connector.update(zli);
    }

    delete(id: ID) {
        return this.connector.delete(id);
    }
}