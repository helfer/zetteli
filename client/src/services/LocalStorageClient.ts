import uuid from 'uuid';

import { ZetteliClient } from './ZetteliClient';
import { ZetteliType } from '../components/Zetteli';

const LOCAL_STORAGE_KEY = 'Zetteli.zettelis';

// TODO(helfer): how do I keep this in sync with ZetteliType?
interface SerializedZetteli {
    id: string;
    body: string;
    tags: string[];
    datetime: string;
}

export default class LocalStorageClient implements ZetteliClient {
    // XXX reading from and writing to local storage
    // will only work in one tab at a time, because
    // writes are not synced and overwrite current contents.
    // That could be solved very easily, but it's not a goal
    // right now.
    private zettelis: ZetteliType[];

    constructor(private store: Storage, private delay: number = 300) {
        this.pullFromStore();
    }

    createNewZetteli(): Promise<string> {
        const newZetteli = {
            tags: ['log', 'personal'],
            datetime: new Date(),
            body: '',
            id: uuid.v4(), // TODO(helfer): this is just asking for trouble
        };
        this.zettelis = [ ...this.zettelis, newZetteli ];
        this.writeToStore();
        return Promise.resolve(newZetteli.id);
    }

    addZetteli(zli: ZetteliType): Promise<boolean> {
        this.zettelis = [ ...this.zettelis, zli];
        this.writeToStore();
        return Promise.resolve(true);
    }

    deleteZetteli(id: string): Promise<boolean> {
        this.zettelis = this.zettelis.filter( zli => zli.id !== id );
        this.writeToStore();
        return Promise.resolve(true);
    }

    updateZetteli(id: string, data: ZetteliType): Promise<boolean> {
        this.zettelis = this.zettelis.map( zli => {
            if (zli.id === id) {
                return { ...zli, ...data };
            } else {
                return zli;
            }
        });
        this.writeToStore();
        return Promise.resolve(true);
    }

    getZetteli(id: string): Promise<ZetteliType | undefined> {
        this.pullFromStore();
        // TODO: reject when Zetteli cannot be found?
        return Promise.resolve(this.zettelis.find( zli => zli.id === id ));
    }

    getAllZettelis(): Promise<ZetteliType[]> {
        this.pullFromStore();
        // NOTE(helfer): Using a delay here to simulate network roundtrip
        return new Promise( (resolve, reject) => {
            setTimeout(() => { resolve(this.zettelis); }, this.delay);
        });
    }

    private writeToStore(): void {
        this.store.setItem(LOCAL_STORAGE_KEY, JSON.stringify(this.zettelis));
    }

    private readFromStore(): ZetteliType[] {
        return JSON.parse(this.store.getItem(LOCAL_STORAGE_KEY) || '[]')
        .map(this.parseZetteli);
    }

    private parseZetteli(zli: SerializedZetteli): ZetteliType {
        return {
            ...zli,
            datetime: new Date(zli.datetime),
        };
    }

    private pullFromStore(): void {
        this.zettelis = this.readFromStore();
    }
}