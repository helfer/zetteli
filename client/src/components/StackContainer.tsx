import * as React from 'react';
import Mousetrap from 'mousetrap';
import FileSaver from 'file-saver';
import moment from 'moment';

import { ZetteliClient } from '../services/ZetteliClient';
import { ZetteliType } from './Zetteli';
import preventDefault from '../utils/preventDefault';
import Stack from './Stack';

export interface Props {
    client: ZetteliClient;
    // TODO(helfer): Is this the best way of controlling which Zettelis to show?
    filterBy: (z: ZetteliType) => boolean;
}

export interface State {
    loading: boolean;
    zettelis: ZetteliType[];
}

export default class StackContainer extends React.Component<Props, State> {
    static defaultProps = {
        filterBy: () => true,
    };

    state = {
        loading: true,
        zettelis: [],
    };

    mousetrap: MousetrapInstance;

    refetchZettelis = () => {
        // TODO: Because this will only work if the call doesn't take too long.
        return this.props.client.getAllZettelis().then( zettelis => {
            this.setState({ zettelis });
        });
    }

    createNewZetteli = () => {
        this.props.client.createNewZetteli()
        .then(() => this.refetchZettelis());
    }

    updateZetteli = (modifiedZetteli: ZetteliType) => {
        this.props.client.updateZetteli(modifiedZetteli.id, modifiedZetteli);
        // NOTE(helfer): Not refreshing here because the client will notify subscribers
    }

    deleteZetteli = (id: string) => {
        this.props.client.deleteZetteli(id)
        .then(() => this.refetchZettelis());
    }

    // Allows the user to save displayed zettelis to a local file
    downloadZettelis = () => {
        function serializeZetteli(z: ZetteliType): string {
            const DATE_RFC2822 = 'ddd, MM MMM YYYY HH:mm:ss ZZ';
            const time = moment.utc(z.datetime).format(DATE_RFC2822);
            // TODO(helfer): Make this string less ugly
            // NOTE(helfer): Should we strip HTML?
            return `
---
>${time}
[${z.tags.join(' ')}]

${z.body}
`;
        }

        const serializedZettelis = this.state.zettelis.map(serializeZetteli);
        var blob = new Blob(serializedZettelis, {type: 'text/plain;charset=utf-8'});
        FileSaver.saveAs(blob, `heftli-${moment().utc().format()}.txt`);
    }

    componentWillMount() {
        this.refetchZettelis().then( () => {
            this.setState({ loading: false });
        });
        this.props.client.subscribe(this.refetchZettelis);
    }

    componentDidMount() {
        // NOTE(helfer): new Mousetrap() works and will bind to document.
        // tslint:disable-next-line no-any
        this.mousetrap = new Mousetrap(undefined as any);
        this.mousetrap.stopCallback = () => false;
        this.mousetrap.bind('command+u', this.createNewZetteli);
        this.mousetrap.bind(['ctrl+s', 'meta+s'], (e) => {
            preventDefault(e); 
            this.downloadZettelis();
        });

    }
    componentWillUnmount() {
        this.mousetrap.unbind('command+u');
        this.mousetrap.unbind(['ctrl+s', 'meta+s']);
        this.props.client.unsubscribe(this.refetchZettelis);
    }

    render() {
        if (this.state.loading) {
            return (
                <div className="ui active inverted dimmer">
                    <div className="ui text loader">Loading</div>
                </div>
            );
        }
        let filteredZettelis = this.state.zettelis.filter(this.props.filterBy);
        return (
          <Stack
            zettelis={filteredZettelis}
            onUpdate={this.updateZetteli}
            onDelete={this.deleteZetteli}
            onCreate={this.createNewZetteli}
          />
        );
    }
}