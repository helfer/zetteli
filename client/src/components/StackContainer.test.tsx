import * as React from 'react';
import * as enzyme from 'enzyme';
import moment from 'moment';
import FileSaver from 'file-saver';
// import { MemoryRouter } from 'react-router';

import StackContainer, { last2days } from './StackContainer';
import { ZetteliType } from './Zetteli';
import LocalStorageClient from '../services/LocalStorageClient';

describe('StackContainer', () => {
    let client: LocalStorageClient;
    let store: Storage;
    let storage: Map<string, string>;
    let zli: ZetteliType;
    let zli2: ZetteliType;

    beforeEach(() => {
        // TODO(helfer): Make a proper mock of LocalStorageClient
        storage = new Map();
        store = {
            getItem: key => storage.get(key) || null,
            length: storage.size,
            clear: () => { storage.clear(); },
            setItem: (key, item) => storage.set(key, item),
            removeItem: key => storage.delete(key),
            key: i => '', // not implemented
        };

        client = new LocalStorageClient(store);
        zli = {
            id: '1',
            body: 'hello',
            datetime: new Date(1500000000000),
            tags: ['t1', 't2'],
        };
        client.addZetteli(zli);
        zli2 = {
            id: '2',
            body: 'hello2',
            datetime: new Date(1500000005000),
            tags: ['t3', 't4'],
        };
        client.addZetteli(zli2);
    });

    it('renders loading state while waiting', () => {
        const stack = enzyme.shallow(
            <StackContainer stackId={'1'} client={client} />
        );
        expect(stack.text()).toContain('Loading');
    });
    
    // listens to cmd+u and creates a new Zetteli if it's pressed
    // TODO(helfer): that test is actually a bit tricky. Let's keep that for later
    it('listens to command+u and adds a new zetteli if called', () => {
        client.createNewZetteli = jest.fn(() => Promise.resolve([]));
        // NOTE(helfer): Have to mount here to get lifecycle events
        const stack = enzyme.mount(<StackContainer stackId={'1'} client={client} />);
        (stack.instance() as StackContainer).mousetrap.trigger('command+u');
        expect(client.createNewZetteli).toHaveBeenCalled();
        stack.unmount();
        // TODO(helfer): Do the mock functions get removed after every test?
    });

    it('downloadZettelis saves zettelis to file', () => {
        const mockSaveAs = jest.fn();
        FileSaver.saveAs = mockSaveAs;
        // TODO(helfer): Find out why global.Blob is not defined
        // tslint:disable-next-line no-any
        (global as any).Blob = (content: string, options: {}) => ({content, options}); 
        const stack = enzyme.shallow(<StackContainer stackId={'1'} client={client} />);
        stack.setState({ zettelis: [zli, zli2], loading: false });
        stack.update();
        (stack.instance() as StackContainer).downloadZettelis();
        expect(mockSaveAs.mock.calls[0][0]).toMatchSnapshot();
    });

    it('Ctrl+s calls downloadZettelis', () => {
        const stack = enzyme.mount(<StackContainer stackId={'1'} client={client} />);
        const mockDownload = jest.fn();
        (stack.instance() as StackContainer).downloadZettelis = mockDownload;
        // TODO(helfer): We're keeping loading=true here so the Router doesn't render,
        // but we should really figure out how to test stuff with a Router...
        stack.setState({ zettelis: [zli, zli2], loading: true });
        (stack.instance() as StackContainer).mousetrap.trigger('ctrl+s');
        stack.unmount();
        expect(mockDownload).toHaveBeenCalled();
    });

    // deleteZetteli calls client.delete
    it('can delete a zetteli', () => {
        client.deleteZetteli = jest.fn(() => Promise.resolve());
        const stack = enzyme.shallow(<StackContainer stackId={'1'} client={client} />);
        // TODO(helfer): Why can't I just call the prop that was passed?
        (stack.instance() as StackContainer).deleteZetteli('2');
        expect(client.deleteZetteli).toHaveBeenCalledWith('2');
    });
    
    // createNewZetteli calls client.create
    it('can create a zetteli', () => {
        client.createNewZetteli = jest.fn(() => Promise.resolve());
        const stack = enzyme.shallow(<StackContainer stackId={'1'} client={client} />);
        (stack.instance() as StackContainer).createNewZetteli();
        expect(client.createNewZetteli).toHaveBeenCalled();
    });
    // updateZetteli calls client.update
    it('can update a zetteli', () => {
        client.updateZetteli = jest.fn(() => Promise.resolve());
        const stack = enzyme.shallow(<StackContainer stackId={'1'} client={client} />);
        (stack.instance() as StackContainer).updateZetteli({ ...zli2, id: zli.id });
        expect(client.updateZetteli).toHaveBeenCalledWith(zli.id, { ...zli2, id: zli.id });
    });

    // refetch calls getAllZettelis
    it('can refetch zettelis', () => {
        client.getAllZettelis = jest.fn(() => Promise.resolve([]));
        const stack = enzyme.shallow(<StackContainer stackId={'1'} client={client} />);
        (stack.instance() as StackContainer).refetchZettelis();
        expect(client.getAllZettelis).toHaveBeenCalled();
    });

    it('last2days filter works correctly', () => {
        // It should include all zettelis from today and yesterday,
        // so at least 24 hours and at most 48 hours.
        zli.datetime = new Date();
        expect(last2days(zli)).toBe(true);
    
        zli.datetime = moment().subtract(1, 'day').toDate();
        expect(last2days(zli)).toBe(true);
    
        zli.datetime = moment().subtract(2, 'days').toDate();
        expect(last2days(zli)).toBe(false);
      });
    
});