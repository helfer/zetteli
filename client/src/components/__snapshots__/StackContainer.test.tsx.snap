// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackContainer downloadZettelis saves zettelis to file 1`] = `
Object {
  "content": Array [
    "
---
>Fri, 07 Jul 2017 02:40:00 +0000
[t1 t2]

hello
",
    "
---
>Fri, 07 Jul 2017 02:40:05 +0000
[t3 t4]

hello2
",
  ],
  "options": Object {
    "type": "text/plain;charset=utf-8",
  },
}
`;
